### DISCUSSION TITLE
Budibase SSO with OIDC (Cognito): 403 Unauthorized

### DISCUSSION THREAD

**Discussion by alexmelissas** [Budibase/budibase#7084] [Help] [answered]
Hello! My SSO setup with OpenID Connect (specifically, Cognito) recently broke with no changes on any end.

I keep getting a 403 Unauthorised, the error on the url is  error_description=invalid_scope&error=invalid_request.

What I noticed is that budibase sends a koa:sess: cookie in the request, including "scope": "openid profile email offline_access".
I don't have offline_access as a scope on my Cognito app integration setup. 

Is this a change that was made recently? Is there a way I can remove that scope from the request in budibase?

**Comment by andyburgessmd** [comment]
I think this is similar to an issue I have with connecting to OneLogin for OIDC. I'm getting the same response and can see the same error.

**Comment by deanhannigan** [comment]
Hi @alexmelissas, @andyburgessmd, thanks for getting in touch!

The "offline_access" scope was added recently to ensure that vendors supply a "refresh_token" during authentication. Without it the new access token bindings would become stale.

@alexmelissas - With cognito, I believe "offline_access" would need to be added to the configured scopes for the identity provider to ensure that a refresh token is supplied.
However, as you point out, the "offline_access" scope would still need to be removed from initial request to cognito itself. This will require a code change on our part to allow auth scopes to be editable.

@andyburgessmd  - With OneLogin, "offline_access" should be a valid scope. In their documentation they state that "Using this scope with Implicit or Authorization Code flow will cause an error." Was this the case for you?

**Reply by andyburgessmd** [reply]
Hi @deanhannigan yes I agree the scope should be valid but if it it not that then there is some other scope that is not valid.
I don't know what flow budibase been developed with, I have configured it as in the docs and don't know how to debug further.
The full utl where I see the error is (with domains changed)...
`https://mydomain.budibase.app/api/global/auth/myname/oidc/callback?error=invalid_scope&error_description=some%20of%20requested%20scopes%20are%20not%20whitelisted&scope=openid&scope=profile&scope=email&scope=offline_access&state=%7B%22handle%22%3A%22oonOamKi3nyRR5bi4zCQ743D%22%7D&ol_oidc_issuer_url=https%3A%2F%2Fmydomain.onelogin.com%2Foidc%2F2`
It could be that the issue is with the scopes parameter being used multiple times in the request to OneLogin instead of just once with all of the scopes. I am using hosted budibase so cannot make changes to test this.

**Reply by alexmelissas** [reply]
Thanks for the clarification @deanhannigan! Any plans/ETA for such a change to be made, to allow editing the auth scopes? 

Current setup renders bb fairly unusable for our use case unfortunately. Cheers!

**Comment by andyburgessmd** [comment]
I have asked OneLogin about this and there response is

JWTs generated from OIDC flows are stateless tokens. Also, v2 password grant requires offline_access scope to be passed if you want the password grant to return a refresh token. 
So in a strict implementation you would require offline_access scope (along with consent prompt, see spec https://openid.net/specs/openid-connect-core-1_0.html#OfflineAccess) in order to return Refresh Token. 
However, we don’t implement offline access in most grants, because we don’t have consents yet. So we just return the RT along with the AT when the token response_type is requested for the code flow. The use of Refresh Tokens is not exclusive to the offline_access use case. The Authorization Server MAY grant Refresh Tokens in other contexts that are beyond the scope of this specification.
In summary as per OIDC Authorization Code document "The “offline_access” scope is not supported on this grant type and will return an error if provided."
https://developers.onelogin.com/openid-connect/api/authorization-code


Can anyone who worked on the OIDC implementation at Budibase respond on this? And hopefully we can get a fix as this is hindering my rollout of an application.

**Reply by andyburgessmd** [reply]
Can confirm the fix that has just gone live does work for OneLogin once the offline_access scope has been removed.

**Comment by andyburgessmd** [comment]
I've raised this as a bug now https://github.com/Budibase/budibase/issues/7267 if anyone wants to follow that.

**Comment by deanhannigan** [comment]
@alexmelissas @andyburgessmd I'll be updating the OIDC config to allow customisable auth scopes. This should hopefully alleviate the issues that your are facing.

**Reply by andyburgessmd** [reply]
Thanks @alexmelissas that's great.

**Comment by laptrc** [comment]
I'm getting the same issue for Slack. Auth scopes customization would be great. 🙏

**Comment by shogunpurple** [comment, marked as answer]
Configurable scopes have now been shipped in both self host and cloud - this should solve the issues in this discussion. Please let us know if anything else crops up.

**Reply by alexmelissas** [reply]
Thank you! All good on my end now.  Thanks @deanhannigan
